packages/toi/index.ts

```diff
@@ -647,69 +647,115 @@ export namespace obj {
     );
 
   /**
-   * Checks that the object obeys a strict structure of properties and property-level validators.
+   * Checks that the object obeys a strict structure of properties and property-level validators. The structure needs to define all properties in a plain JavaScript object and its prototype chain is not inspected.
    *
    * It is a strict 1:1 validation. Keys not found on the value will trigger a {@link ValidationError},
-   * including keys found in the `value` but not in the structure.
+   * including keys found in the `value` but not in the structure. This excludes non-enumerable properties (like `toString` or other built-in functions) or properties defined on symbols.
+   * You can disable this strict 1:1 mapping by setting the `lenient` option to `true.
    *
-   * The `value` is never returned as-is, but is made a copy of with the own properties according to
+   * The prototype chain of the passed value will be walked and inherited properties will be validated. You can disable this by setting the `own` option to `true`.
+   *
+   * The `value` is never returned as-is, but is made a copy of with the properties according to
    * the defined `structure`. Therefore, the property-level validators may transform the property
    * values. This safeguards from `instanceof` checks, including prototype-hijacking and similar
-   * issues.
+   * issues. The returned value will always have the Object prototype, as the prototype chain is not copied.
    *
    * @param structure the definition structure of the object
-   * @param options the validation options, like optional (missing) fields
+   * @param options the validation options
    */
-  export const keys = <X extends object, Y>(
+  export const keys = <X extends object, Y, M extends Y>(
     structure: { [K in keyof Y]: Validator<any, Y[K]> },
-    options?: { missing?: (keyof Y)[] }
+    options?: {
+      /** Set the keys that may be missing in the value. */
+      missing?: (keyof M)[];
+      /** Only validate the value's own properties and ignore inherited properties. */
+      own?: boolean;
+      /** Don't validate that value matches 1:1 with the structure. */
+      lenient?: boolean;
+    }
   ) => {
-    const missing: { [key in keyof Y]?: true } = {};
+    const missing: { [key in keyof M]?: true } = {};
 
     if (options && options.missing && options.missing.length > 0) {
       for (let i = 0; i < options.missing.length; i += 1) {
         missing[options.missing[i]] = true;
       }
     }
 
+    const walkPrototype = !options || !options.own;
+
     return wrap<X, Y>("obj.keys", value => {
       if (null === value || undefined === value) {
         return value;
       }
 
-      let reasons: {
+      let hasReasons = false;
```

Comment:
@hf Did you add this flag to get rid of managing the initialization of the `reasons` object on multiple places? But couldn't you perform the check for reasons in another way, where you would depend on the `reasons` object only?

Reply:
Agreeing with @gphfour. This is a bug waiting to happen.

Reply:
The `hasReasons` is there because there's no efficient way to check if an object is empty.

Reply:
And if I leave it as it was before with `null` I have to write a bunch more `if` statements which are difficult to cover with tests.

Reply:
https://stackoverflow.com/questions/679915/how-do-i-test-for-an-empty-javascript-object

```javascript
Object.keys(reasons).length
```
... is bad because it materializes all properties in an array.

```javascript
for (let key in reasons) {
  hasReasons = true;
  break;
}
```
... does not seem like an improvement.